[RFC] Fix a few uninitialized variable warnings.

jszakmeister opened the pull request:
Fixes a few more warnings in release mode.

jszakmeister committed:
Fix a couple uninitialized variable warnings in the release build.

Reviewed-by: Eliseo Martínez <eliseomarmol@gmail.com>
Reviewed-by: Marco Hinz <mh.codebro@gmail.com>
Reviewed-by: Björn Linse <bjorn.linse@gmail.com>

mhinz commented:
LGTM

bfredl commented:
LGTM :+1:

elmart commented:
I've noticed we can be entering a vicious circle with these changes.

In the past, I've fixed a good bunch of "dead store" clang analysis warnings. Those warnings are fired when you assign something to a variable, that never gets used. That occurs specially often with code that initializes variables within conditionals in a complicated, but also initializes variables at declaration time.

In such cases, it's better to leave the variables unitialized, cause that way the compiler would warn if code modifications make possible a new path that doesn't correctly initialize the variable. Also, they get rid of the clang anlysis warning.

Now, when initialization logic is complex, it's sometimes necessary to convince the compiler/analyzer the variable is always being initialized, through using asserts.

But then, when we remove asserts in Release build, compiler complains again.

And, if we initialize the variable to avoid compiler warning, then we get back the clang analyzer warning on dead store for the regular build with asserts.

So, if possible, these kind of problems with Release build should be tried to be solved without initializing variables. When that is not possible, we must choose between compiler warning on Release (unsupported) build and clang-analysis warning on regular (supported) build. It's obvious which one I do prefer. ;-)

I'll give it a try to see if above commits can be fixed other way.

elmart commented on src/nvim/hardcopy.c:
```diff
@@ -2117,7 +2117,7 @@ int mch_print_init(prt_settings_T *psettings, char_u *jobname, int forceit)
   if (!(props & ENC_8BIT) && ((*p_pmcs != NUL) || !(props & ENC_UNICODE))) {
     p_mbenc_first = NULL;
-    int effective_cmap;
+    int effective_cmap = 0;
```
Instead of this, see below comment.

elmart replied:
Argg, I can't comment on unchanged line.
Well, on line 2133, changing order of conditions from `if (p_mbchar == NULL && p_mbenc_first != NULL) {` to the opposite, clears the warning without the variable initialization.

jszakmeister replied:
I like that.  I'll fix it up later tonight.

elmart replied:
I'm afraid you've forgotten to remove effective_cmap initialization.

jszakmeister replied:
Fixed.  Thanks!

elmart commented on src/nvim/msgpack_rpc/server.c:
```diff
@@ -224,7 +224,7 @@ int server_start(const char *endpoint)
 void server_stop(char *endpoint)
 {
-  Server *server;
+  Server *server = NULL;
```
This one is a bug, in fact. One that would never occur, but a bug after all. Problem is in:
```
  int i = 0;  // The index of the server whose address equals addr.
  for (; i < servers.ga_len; i++) {
    server = ((Server **)servers.ga_data)[i];
    if (strcmp(addr, server->addr) == 0) {
      break;
    }
  }

  if (i == servers.ga_len) {
    ELOG("Not listening on %s", addr);
    return;
  }
```
This code wouldn't work for server.ga_len < 0.
We know that won't happen, but from the compiler point of view, it's an int, and it could.
That case leads to function not exiting at ELOG block, and server being used without initialization.
It can all be fixed (without initializing variable) by changing it into:
```
  int i = -1;  // The index of the server whose address equals addr.                                                                                  
   for (int j = 0; j < servers.ga_len; j++) {                                                                                                          
     server = ((Server **)servers.ga_data)[j];                                                                                                         
     if (strcmp(addr, server->addr) == 0) {                                                                                                            
       i = j;                                                                                                                                          
       break;                                                                                                                                          
     }                                                                                                                                                 
   }                                                                                                                                                   
                                                                                                                                                       
   if (i == -1) {                                                                                                                                      
     ELOG("Not listening on %s", addr);                                                                                                                
     return;                                                                                                                                           
   }
```
Or, alternatively, introducing a server_found bool variable.
This is an example of how leaving variables uninitialized where it makes sense can actually help.

jszakmeister replied:
Hrmph.  I had assumed that `ga_len` was an unsigned type and didn't even think to look. :-(

I think the fix is far simpler here:

```diff
diff --git a/src/nvim/msgpack_rpc/server.c b/src/nvim/msgpack_rpc/server.c
index 8fb2902..5e3fd10 100644
--- a/src/nvim/msgpack_rpc/server.c
+++ b/src/nvim/msgpack_rpc/server.c
@@ -238,7 +238,7 @@ void server_stop(char *endpoint)
     }
   }

-  if (i == servers.ga_len) {
+  if (i >= servers.ga_len) {
     ELOG("Not listening on %s", addr);
     return;
   }
```

Thanks for looking at it closer though!

elmart replied:
:+1:

elmart replied:
In fact, refactoring ga_len to `size_t` is something we should do at some point. Apart from causing subtle bugs like this, it's also the main cause for a lot of casting between int and size_t. It's something I have in my type-refactoring-todo list, but I haven't reached it yet.

elmart commented on src/nvim/eval.c:
```diff
@@ -18911,7 +18911,7 @@ call_user_func (
   char_u      *name;
   proftime_T wait_start;
-  proftime_T call_start;
+  proftime_T call_start = 0;
```
This one is a bit harder, but solvable nonetheless.
Both the initialization of call_start:
```
   if (do_profiling == PROF_YES) {
    if (!fp->uf_profiling && has_profiling(FALSE, fp->uf_name, NULL))
      func_do_profile(fp);
    if (fp->uf_profiling
        || (fc->caller != NULL && fc->caller->func->uf_profiling)) {
      ++fp->uf_tm_count;
      call_start = profile_start();
      fp->uf_tm_children = profile_zero();
    }
    script_prof_save(&wait_start);
  }
```
and the (supposedly offending) usage place:
```
  if (do_profiling == PROF_YES && (fp->uf_profiling
                                   || (fc->caller != NULL &&
                                       fc->caller->func->uf_profiling))) {
    call_start = profile_end(call_start);
    ...
   }
```
are under the same logical condition.
So, they should both execute, or none, but it seems the compiler can't follow that implication.
Therefore, solution is a bit of refactoring to make sure compiler understands condition is the same at both points.
For example, transforming the first block into:
```
  bool func_not_yet_profiling_but_should =                                                                                                          
     do_profiling == PROF_YES                                                                                                                        
     && !fp->uf_profiling && has_profiling(FALSE, fp->uf_name, NULL);                                                                                
                                                                                                                                                     
   if (func_not_yet_profiling_but_should)                                                                                                            
     func_do_profile(fp);                                                                                                                            
                                                                                                                                                     
   bool func_or_func_caller_profiling =                                                                                                              
     do_profiling == PROF_YES                                                                                                                        
     && (fp->uf_profiling                                                                                                                            
         || (fc->caller != NULL && fc->caller->func->uf_profiling));                                                                                 
                                                                                                                                                     
   if (func_or_func_caller_profiling) {                                                                                                              
     ++fp->uf_tm_count;                                                                                                                              
     call_start = profile_start();                                                                                                                   
     fp->uf_tm_children = profile_zero();                                                                                                            
   }                                                                                                                                                 
                                                                                                                                                     
   if (do_profiling == PROF_YES) {                                                                                                                   
     script_prof_save(&wait_start);                                                                                                                  
   }
```
which is also a bit easier to read, and the second block into
```
  if (func_or_func_caller_profiling) {                                                                                                              
     call_start = profile_end(call_start);
     ...
  }
```
makes the warning disappear without initializing variables.
I'm assuming that value of conditions can't change in-between the two point, which seems a reasonable assumption to me (it it did, it'd probably be a bug rather than an intended purpose).

elmart commented:
Here you have a patch with all the changes I suggested:
```
diff --git a/src/nvim/eval.c b/src/nvim/eval.c
index d6a8351..1392a12 100644
--- a/src/nvim/eval.c
+++ b/src/nvim/eval.c
@@ -19086,15 +19086,26 @@ call_user_func (
       --no_wait_return;
     }
   }
+
+  bool func_not_yet_profiling_but_should =
+    do_profiling == PROF_YES
+    && !fp->uf_profiling && has_profiling(FALSE, fp->uf_name, NULL);
+
+  if (func_not_yet_profiling_but_should)
+    func_do_profile(fp);
+
+  bool func_or_func_caller_profiling =
+    do_profiling == PROF_YES
+    && (fp->uf_profiling
+        || (fc->caller != NULL && fc->caller->func->uf_profiling));
+
+  if (func_or_func_caller_profiling) {
+    ++fp->uf_tm_count;
+    call_start = profile_start();
+    fp->uf_tm_children = profile_zero();
+  }
+
   if (do_profiling == PROF_YES) {
-    if (!fp->uf_profiling && has_profiling(FALSE, fp->uf_name, NULL))
-      func_do_profile(fp);
-    if (fp->uf_profiling
-        || (fc->caller != NULL && fc->caller->func->uf_profiling)) {
-      ++fp->uf_tm_count;
-      call_start = profile_start();
-      fp->uf_tm_children = profile_zero();
-    }
     script_prof_save(&wait_start);
   }
 
@@ -19117,9 +19128,7 @@ call_user_func (
     rettv->vval.v_number = -1;
   }
 
-  if (do_profiling == PROF_YES && (fp->uf_profiling
-                                   || (fc->caller != NULL &&
-                                       fc->caller->func->uf_profiling))) {
+  if (func_or_func_caller_profiling) {
     call_start = profile_end(call_start);
     call_start = profile_sub_wait(wait_start, call_start);
     fp->uf_tm_total = profile_add(fp->uf_tm_total, call_start);
diff --git a/src/nvim/hardcopy.c b/src/nvim/hardcopy.c
index 6e9ad02..0cbbe95 100644
--- a/src/nvim/hardcopy.c
+++ b/src/nvim/hardcopy.c
@@ -2130,7 +2130,7 @@ int mch_print_init(prt_settings_T *psettings, char_u *jobname, int forceit)
       }
 
     /* Use first encoding matched if no charset matched */
-    if (p_mbchar == NULL && p_mbenc_first != NULL) {
+    if (p_mbenc_first != NULL && p_mbchar == NULL) {
       p_mbenc = p_mbenc_first;
       cmap = effective_cmap;
     }
diff --git a/src/nvim/msgpack_rpc/server.c b/src/nvim/msgpack_rpc/server.c
index 8fb2902..d526cfe 100644
--- a/src/nvim/msgpack_rpc/server.c
+++ b/src/nvim/msgpack_rpc/server.c
@@ -230,15 +230,16 @@ void server_stop(char *endpoint)
   // Trim to `ADDRESS_MAX_SIZE`
   xstrlcpy(addr, endpoint, sizeof(addr));
 
-  int i = 0;  // The index of the server whose address equals addr.
-  for (; i < servers.ga_len; i++) {
-    server = ((Server **)servers.ga_data)[i];
+  int i = -1;  // The index of the server whose address equals addr.
+  for (int j = 0; j < servers.ga_len; j++) {
+    server = ((Server **)servers.ga_data)[j];
     if (strcmp(addr, server->addr) == 0) {
+      i = j;
       break;
     }
   }
 
-  if (i == servers.ga_len) {
+  if (i == -1) {
     ELOG("Not listening on %s", addr);
     return;
   }
```

elmart commented:
Previous patch is to apply on master, not this branch.

jszakmeister commented:
> I've noticed we can be entering a vicious circle with these changes.
> 
> In the past, I've fixed a good bunch of "dead store" clang analysis warnings. Those warnings are fired when you assign something to a variable, that never gets used. That occurs specially often with code that initializes variables within conditionals in a complicated way, but also initializes them at declaration time, usually just out of habit.
> 
> In such cases, it's better to leave the variables unitialized, cause that way the compiler would warn if code modifications make possible a new path that doesn't correctly initialize the variable. Also, they get rid of the clang anlaysis warning.
> 
> Now, when initialization logic is complex, it's sometimes necessary to convince the compiler/analyzer the variable is always being initialized, through using asserts.
> 
> But then, when we remove asserts in Release build, compiler complains again.
> 
> And, if we initialize the variable to avoid compiler warning, then we get back the clang analyzer warning on dead store for the regular build with asserts.
> 
> So, if possible, these kind of problems with Release build should be tried to be solved without initializing variables. When that is not possible, we must choose between compiler warning on Release (unsupported) build and clang-analysis warning on regular (supported) build. It's obvious which one I do prefer. ;-)

Not to add to the mix, but sometimes compilers get it wrong too (the can't take the analysis far enough to determine it correctly).  GCC has gotten me a few times on that in the past on other projects. :-(

Release is not an unsupported build, it's just not the default. ;-)  Quite honestly, I would loved to have had a build as part of Travis CI long ago, but the issues kept popping up faster than I could fix them.

Thank you for taking a deeper look at this @elmart.  I've incorporated your suggestion for the `effective_cmap` fix.  I split the `call_start` fix into a separate commit and put you as the author.  I imagine it took a bit of time to make it all happy.  Thanks for that.

elmart committed:
Fix an uninitialized variable warning for call_start.

elmart commented:
> sometimes compilers get it wrong too (the can't take the analysis far enough to determine it correctly). GCC has gotten me a few times

Yes. In fact, none of these warning were fired by clang. Only by gcc. Not sure, though, if that's because clang compiler analysis is is better, or because it doesn't try to detect such things. I tend to think the former, but not sure.

elmart commented:
Maybe PR title should be changed now? It doesn't initialize variables anymore.

jszakmeister commented:
> Yes. In fact, none of these warning were fired by clang. Only by gcc. Not sure, though, if that's because clang compiler analysis is is better, or because it doesn't try to detect such things. I tend to think the former, but not sure.

/me agrees.

I removed the effective_cmap initialization and will let it run through the tests again.  Thanks again for all your time and sorry for all the back and forth!

elmart commented:
You're welcome. We're here to help when possible.

LGTM.

elmart commented:
There are other PR's stalling at Mac build for gcc. Travis status say they're ok, but maybe they just don't know yet.

jszakmeister commented:
It was actually failing before... It looked like something was going wrong with the profiling (there was an error trying to merge a .gcda file and profiling statements were appearing in the output).  It seemed odd, so I assumed it was a transient issue and restarted the build. :-/  I ran the build on my Linux VM and my laptop (using OS X and gcc 4.9) with gcov turned on, but didn't see any issues on either one.

splinterofchaos commented:
LGTM :+1:

jszakmeister commented:
Thanks everyone for the review and comments!  Back to the conference...
